File: Remote/Source/Networking/NetworkDevice.m
//
// NetworkDevice.m
// Remote
//
// Created by Jason Cardwell on 9/25/12.
// Copyright (c) 2012 Moondeer Studios. All rights reserved.
//

#import "NetworkDevice.h"
#import "CoreDataManager.h"
#import "NDiTachDevice.h"
#import "ISYDevice.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = LOG_CONTEXT_CONSOLE;
#pragma unused(ddLogLevel,msLogContext)


////////////////////////////////////////////////////////////////////////////////
#pragma mark - Abstract
////////////////////////////////////////////////////////////////////////////////


@interface NetworkDevice ()
@property (nonatomic, copy, readwrite) NSString * uniqueIdentifier;
@end

@implementation NetworkDevice

@dynamic uniqueIdentifier, componentDevices;

/// deviceExistsWithDeviceUUID:
/// @param identifier description
/// @return BOOL
+ (BOOL)deviceExistsWithUniqueIdentifier:(NSString *)identifier {
  return [self countOfObjectsWithPredicate:NSPredicateMake(@"uniqueIdentifer == %@", identifier)] > 0;
}

/// importObjectFromData:context:
/// @param data description
/// @param moc description
/// @return instancetype
+ (instancetype)importObjectFromData:(NSDictionary *)data context:(NSManagedObjectContext *)moc {
  if (self == [NetworkDevice class] && [@"itach" isEqualToString:data[@"type"]])
    return [NDiTachDevice importObjectFromData:data context:moc];
  else if (self == [NetworkDevice class] && [@"isy" isEqualToString:data[@"type"]])
    return [ISYDevice importObjectFromData:data context:moc];
  else
    return [super importObjectFromData:data context:moc];
}

/// updateWithData:
/// @param data description
- (void)updateWithData:(NSDictionary *)data {

  [super updateWithData:data];

  self.uniqueIdentifier = data[@"unique-identifier"];
  self.name             = data[@"name"];

}

/// networkDeviceForBeaconData:context:
/// @param message description
/// @param moc description
/// @return instancetype
+ (instancetype)networkDeviceFromDiscoveryBeacon:(NSString *)message context:(NSManagedObjectContext *)moc {
  return nil;
}

/// JSONDictionary
/// @return MSDictionary *
- (MSDictionary *)JSONDictionary {

  MSDictionary * dictionary = [super JSONDictionary];

  SafeSetValueForKey(self.name,             @"name",              dictionary);
  SafeSetValueForKey(self.uniqueIdentifier, @"unique-identifier", dictionary);


  return dictionary;

}

- (NSString *)multicastGroupAddress { return nil; }
- (NSString *)multicastGroupPort    { return nil; }

@end

File: Remote/Source/Bank/Bankables/NetworkDevice.swift
//
//  NetworkDevice.swift
//  Remote
//
//  Created by Jason Cardwell on 9/30/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(NetworkDevice)
class NetworkDevice: BankableModelObject {

  @NSManaged var uniqueIdentifier: String!
  @NSManaged var componentDevices: NSSet?

  class func deviceExistsWithIdentifier(identifier: String) -> Bool {
    return countOfObjectsMatchingPredicate(∀"uniqueIdentifier == \"\(identifier)\"", context: DataManager.rootContext) > 0
  }

  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)
    uniqueIdentifier = data["unique-identifier"] as? String ?? uniqueIdentifier
  }

  /**
  fetchOrImportObjectWithData:context:

  :param: data [String:AnyObject]
  :param: context NSManagedObjectContext

  :returns: NetworkDevice?
  */
  override class func fetchOrImportObjectWithData(data: [String:AnyObject], context: NSManagedObjectContext) -> NetworkDevice? {

    var device: NetworkDevice?

    // Try getting the type of device to import
    if let type = data["type"] as? String {

      var entityName: String?
      var deviceType: NetworkDevice.Type = NetworkDevice.self

      // Import with parameters derived from specified type
      switch type {
        case "itach":
          device = importObjectForEntity("ITachDevice", forType: ITachDevice.self, fromData: data, context: context) as? NetworkDevice
        case "isy":
          device = importObjectForEntity("ISYDevice", forType: ISYDevice.self, fromData: data, context: context) as? NetworkDevice
        default:
          break
      }
    }

    return device
  }
  class var rootCategory: Bank.RootCategory {
    let networkDevices = findAllSortedBy("name", ascending: true, context: DataManager.rootContext) as? [NetworkDevice]
    return Bank.RootCategory(label: "Network Devices",
                             icon: UIImage(named: "937-wifi-signal")!,
                             items: networkDevices ?? [],
                             editableItems: true)
  }

}

extension NetworkDevice: MSJSONExport {

  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()
      safeSetValue(uniqueIdentifier, forKey: "unique-identifier", inDictionary: dictionary)
      dictionary.compact()
      dictionary.compress()
      return dictionary
  }

}
